src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala

```diff
@@ -548,8 +548,8 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
       0 until u2() foreach (_ => parseMethod())
       val needsConstructor = (
            !sawPrivateConstructor
-        && !(instanceScope containsName nme.CONSTRUCTOR)
-        && ((sflags & INTERFACE) == 0 || (sflags | JAVA_ANNOTATION) != 0)
+        && !instanceScope.containsName(nme.CONSTRUCTOR)
+        && ((sflags & (INTERFACE|JAVA_ANNOTATION)) == (INTERFACE|JAVA_ANNOTATION))
```

Comment:
`val needsConstructor = sflags & JAVA_ANNOTATION) != 0` and removing everything else (`sawPrivateConstructor` / `instanceScope`) doesn't work?

Reply:
I was being conservative out of an abundance of fear.

Reply:
I have forgotten the correct answer here.